Denis_Protsenko.py

```diff
@@ -0,0 +1,218 @@
+
+    def __hash__(self):
+        return hash((self.title, self.author, self.year))
+
+
+# Клас журналу
+class Magazine(PublicModel):
+    def __init__(self, magazine_model: MagazineModel):
+        self.title = magazine_model.title
+        self.author = magazine_model.author
+        self.year = magazine_model.year
+
+    def book_info(self):  # Implementing the abstract method
+        return f"{self.title}, Автор: {self.author}, Рік видання: {self.year}"
+
+    def __eq__(self, other):
+        if not isinstance(other, Magazine):
+            return False
+        return (self.title, self.author, self.year) == (other.title, other.author, other.year)
+
+    def __hash__(self):
+        return hash((self.title, self.author, self.year))
+
+
+# Декоратор для логування при додаванні нової книги до бібліотеки
+def log_addition(func):
+    def wrapper(self, book_model):
+        if isinstance(book_model, BookModel):
+            print(f"Додаємо: {book_model.title} (Автор: {book_model.author})")
+        elif isinstance(book_model, MagazineModel):
+            print(f"Додаємо: {book_model.title} (Автор: {book_model.author})")
+        return func(self, book_model)
+    return wrapper
+
+
+# Декоратор, який перевіряє наявність книги в бібліотеці перед її видаленням
+def check_book_existence(func):
+    def wrapper(*args, **kwargs):
+        library, _book = args[0], args[1]
+        if _book in library.books:
+            removed_book = func(*args, **kwargs)
+            if removed_book:
+                print(f"\nВидаленo: {removed_book}")
+                return removed_book  # Возвращаем результат удаления
+        else:
+            print(f"\nКнига не знайдена у бібліотеці.")
+            return None  # Возвращаем None, если книга не была найдена
+    return wrapper
+
+
+# Клас бібліотеки
+class Library:
+    def __init__(self):
+        self.books = set()
+        self.books_by_author_dict = defaultdict(list)  # Словник автор: [список книг]
+        print("Бібліотека створена\n")
+
+    # Додавання книги в бібліотеку
+    @log_addition
+    def add_book(self, book_model: Union[BookModel, MagazineModel]):
+        if isinstance(book_model, BookModel):
+            book = Book(book_model)
+        elif isinstance(book_model, MagazineModel):
+            existing_magazine = next((book for book in self.books if book.title == book_model.title and isinstance(book, Magazine)), None)
+            if existing_magazine:
+                print(f"Журнал {book_model.title} вже є у бібліотеці.")
+                return
+            book = Magazine(book_model)
+        else:
+            return
+
+        if book not in self.books:
+            self.books.add(book)
+            self.books_by_author_dict[book.author].append(book)
+
+
+    # Виведення книг по імені автора
+    @staticmethod
+    def print_books_by_author(author_name, books_by_author_dict):
+        if author_name in books_by_author_dict:
+            books = books_by_author_dict[author_name]
```

Comment:
Interesting solution with creating a dict which will store all author books by key.
It really will work much faster than simple loop with checking attributes for each book.
But it can increase a memory usage, because you'll same same books but in another variable.
In real life books will be stored in database, and you'll be able to filter books by author, and even add an index to this field.
Now is late night, so I'm not able to propose best solution. But I like you idea, and it can be discussed later